pulsar/pulsar.sh

```diff
@@ -0,0 +1,225 @@
+      sleep "${sleep_time}"
+    fi
+  done
+
+  if ! ((installation_succeeded)); then
+    echo "Final attempt of '${cmd[*]}'..."
+    # Let any final error propagate all the way out to any error traps.
+    "${cmd[@]}"
+  fi
+}
+
+function edit() {
+  if [ "$#" == "3" ]; then
+   grep -q "$1" "$3" && sed -i "s/"$1"/"$2"/" "$3" || echo "$2" >> "$3"
+  else
+   echo "$1" >> "$2"
+  fi
+}
+
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+  return 1
+}
+
+function retry_apt_command() {
+  cmd="$1"
+  for ((i = 0; i < 10; i++)); do
+    if eval "$cmd"; then
+      return 0
+    fi
+    sleep 5
+  done
+  return 1
+}
+
+function update_apt_get() {
+  retry_apt_command "apt-get update"
+}
+
+
+function fetch_zookeeper_list() {
+  local zookeeper_client_port
+  zookeeper_client_port=$(grep 'clientPort' /etc/zookeeper/conf/zoo.cfg |
+    tail -n 1 |
+    cut -d '=' -f 2)
+
+  local zookeeper_list
+  zookeeper_list=$(grep '^server\.' /etc/zookeeper/conf/zoo.cfg |
+    cut -d '=' -f 2 |
+    cut -d ':' -f 1 |
+    sort |
+    uniq |
+    sed "s/$/:${zookeeper_client_port}/" |
+    xargs echo |
+    sed "s/ /,/g")
+    
+    ZOOKEEPER_ADDRESSES="${zookeeper_list}"
+
+  # If attempt fails, error out.
+  if [[ -z "${zookeeper_list}" ]]; then
+    err 'Failed to find configured Zookeeper list. Please remember to include Zookeeper optional component'
+  fi
+  
+  ZK_ADDRS="${zookeeper_list}"
+  ZK_ADDRS=${ZK_ADDRS//,/$'\n'}  
+  ZK_ADDRS=($ZK_ADDRS)
+
+  for i in "${!ZK_ADDRS[@]}"; do
+    local serverNo=$i
+    serverNo=$((serverNo+1))
+    edit "server.${serverNo}=${ZK_ADDRS[$i]}" conf/zookeeper.conf
+    done
+
+  ZOOKEEPER_ADDRESS="${zookeeper_list%%,*}"
+}
+
+function fetch_urls() {
+  if [ "$NUM_WORKERS" == "0" ]; then
+    WEB_SERVICE_URL+="$CLUSTER_NAME-w:$WEB_SERVICE_URL_PORT"
+    WEB_SERVICE_URL_TLS+="$CLUSTER_NAME-w:$WEB_SERVICE_URL_TLS_PORT"
+    BROKER_SERVICE_URL+="$CLUSTER_NAME-w:$BROKER_SERVICE_URL_PORT"
+    BROKER_SERVICE_URL_TLS+="$CLUSTER_NAME-w:$BROKER_SERVICE_URL_TLS_PORT"
+  else
+    for ((i=0; i<$NUM_WORKERS; i++))
+    do
+      if [ "$i" == $(($NUM_WORKERS-1)) ]; then
+        WEB_SERVICE_URL+="$CLUSTER_NAME-w-$i:$WEB_SERVICE_URL_PORT"
+        WEB_SERVICE_URL_TLS+="$CLUSTER_NAME-w-$i:$WEB_SERVICE_URL_TLS_PORT"
+        BROKER_SERVICE_URL+="$CLUSTER_NAME-w-$i:$BROKER_SERVICE_URL_PORT"
+        BROKER_SERVICE_URL_TLS+="$CLUSTER_NAME-w-$i:$BROKER_SERVICE_URL_TLS_PORT"
+      else 
+      WEB_SERVICE_URL+="$CLUSTER_NAME-w-$i:$WEB_SERVICE_URL_PORT,"
+      WEB_SERVICE_URL_TLS+="$CLUSTER_NAME-w-$i:$WEB_SERVICE_URL_TLS_PORT,"
+      BROKER_SERVICE_URL+="$CLUSTER_NAME-w-$i:$BROKER_SERVICE_URL_PORT,"
+      BROKER_SERVICE_URL_TLS+="$CLUSTER_NAME-w-$i:$BROKER_SERVICE_URL_TLS_PORT,"
+      fi
+    done
+  fi
+}
+
+function install_and_configure_pulsar() {
+    retry_command wget https://archive.apache.org/dist/pulsar/pulsar-2.6.0/apache-pulsar-2.6.0-bin.tar.gz
```

Comment:
Can we make Pulsar version configurable with metadata?

Reply:
ack. Done.